doc/doxygen/src/driver-guide.md

```diff
@@ -0,0 +1,350 @@
+# Guide for writing a device driver in RIOT
+This document describes the requirement, design objectives, and some
+non-function details when writing device drivers in/for RIOT. The term device
+driver in this context includes all 'CPU-external' devices connected to the CPU
+typically via peripherals like SPI, I2C, UART, GPIO, and similar. CPU
+peripherals itself are in RIOT not considered to be device drivers, but
+peripheral or low-level drivers. Typical devices are network devices like
+radios, Ethernet adapters, sensors, and actuators.
+
+
+## General Design Objectives
+Device drivers should be as easy to use as possible. This implies an
+'intialize->ready' paradigm, meaning, that device drivers should be ready to use
+and in operation right after they have been initialized. On top, devices should
+work with physical values where ever possible. So e.g. sensors should return
+already converted values in some physical unit instead of RAW data, so that
+users can work directly with data from different devices directly without having
+to care for device specific conversion.
+
+Additionally towards ease of use, all device drivers in RIOT should provide a
+similar 'look and feel'. They should behave similar concerning things like their
+state after initialization, like their used data representation and so on.
+
+Secondly, all device drivers should be optimized for minimal RAM/ROM usage, as
```

Comment:
Note that "exposing all functionalities" doesn't necessarily correlate with "higher RAM/ROM consumption". 
If the interface on top doesn't access a feature on top, then it won't be linked. There are e.g several features implemented in the `sx127x` driver that never get to the binary but they are still there if someone need them.

In fact, I think it's easier if the presentation layer is well defined and whatever goes on top choose the right components. What do you think?

Reply:
I agree, but I would add this in a follow-up PR.

In the end, if a PR adds an "uncommon" feature someone is interested in having and maintaining, and it doesn't add overhead for other users, there is to me no reason to just accept that PR.